pyro/contrib/glmm/guides.py

```diff
@@ -6,35 +6,49 @@
 import pyro.distributions as dist
 from pyro import poutine
-from pyro.contrib.util import tensor_to_dict, rmv, rvv, rtril
+from pyro.contrib.util import (
+    tensor_to_dict, rmv, rvv, rtril, lexpand, iter_plates_to_shape
+)
 from pyro.ops.linalg import rinverse
 
+try:
+    from contextlib import ExitStack  # python 3
+except ImportError:
+    from contextlib2 import ExitStack  # python 2
 
-class LinearModelGuide(nn.Module):
 
-    def __init__(self, d, w_sizes, tikhonov_init=-2., scale_tril_init=3.):
+class LinearModelPosteriorGuide(nn.Module):
+
+    def __init__(self, d, w_sizes, y_sizes, regressor_init=0., scale_tril_init=3., use_softplus=True, **kwargs):
         """
         Guide for linear models. No amortisation happens over designs.
         Amortisation over data is taken care of by analytic formulae for
         linear models (heavy use of truth).
 
-        :param int d: the number of designs
-        :param dict w_sizes: map from variable string names to int.
-        :param float tikhonov_init: initial value for `tikhonov_diag` parameter.
-        :param float scale_tril_init: initial value for `scale_tril` parameter.
+        :param tuple d: the shape by which to expand the guide parameters, e.g. `(num_batches, num_designs)`.
+        :param dict w_sizes: map from variable string names to int, indicating the dimension of each
+                             weight vector in the linear model.
+        :param float regressor_init: initial value for the regressor matrix used to learn the posterior mean.
+        :param float scale_tril_init: initial value for posterior `scale_tril` parameter.
+        :param bool use_softplus: whether to transform the regressor by a softplus transform: useful if the
+                                  regressor should be nonnegative but close to zero.
         """
-        super(LinearModelGuide, self).__init__()
+        super(LinearModelPosteriorGuide, self).__init__()
         # Represent each parameter group as independent Gaussian
         # Making a weak mean-field assumption
         # To avoid this- combine labels
-        self.tikhonov_diag = nn.Parameter(
-                tikhonov_init*torch.ones(sum(w_sizes.values())))
+        if not isinstance(d, (tuple, list, torch.Tensor)):
+            d = (d,)
+        self.regressor = {l: nn.Parameter(
+                regressor_init * torch.ones(*(d + (p, sum(y_sizes.values()))))) for l, p in w_sizes.items()}
         self.scale_tril = {l: nn.Parameter(
-                scale_tril_init*torch.ones(d, p, p)) for l, p in w_sizes.items()}
+                scale_tril_init * lexpand(torch.eye(p), *d)) for l, p in w_sizes.items()}
         # This registers the dict values in pytorch
         # Await new version to use nn.ParamterDict
-        self._registered = nn.ParameterList(self.scale_tril.values())
+        self._registered_re = nn.ParameterList(self.regressor.values())
+        self._registered_st = nn.ParameterList(self.scale_tril.values())
```

Comment:
Do you want to switch to [`torch.nn.ParameterDict`](https://pytorch.org/docs/stable/nn.html#torch.nn.ParameterDict) here now that it's available in 1.1?

Reply:
Oh yes